baseapp/test_helpers.go

```diff
@@ -34,24 +35,34 @@ func (app *BaseApp) SimDeliver(txEncoder sdk.TxEncoder, tx sdk.Tx) (sdk.GasInfo,
 	if err != nil {
 		return sdk.GasInfo{}, nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "%s", err)
 	}
-	gasInfo, result, _, _, err := app.runTx(runTxModeDeliver, bz)
+	gasInfo, result, _, err := app.runTx(execModeFinalize, bz)
+	return gasInfo, result, err
+}
+
+func (app *BaseApp) SimTxFinalizeBlock(txEncoder sdk.TxEncoder, tx sdk.Tx) (sdk.GasInfo, *sdk.Result, error) {
+	// See comment for Check().
+	bz, err := txEncoder(tx)
+	if err != nil {
+		return sdk.GasInfo{}, nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "%s", err)
+	}
+
+	gasInfo, result, _, err := app.runTx(execModeFinalize, bz)
 	return gasInfo, result, err
 }
 
-// Context with current {check, deliver}State of the app used by tests.
-func (app *BaseApp) NewContext(isCheckTx bool, header cmtproto.Header) sdk.Context {
+func (app *BaseApp) NewContext(isCheckTx bool, header cmtproto.Header) sdk.Context { // todo discuss how to remove header, wrapper or no
```

Comment:
should we break this api? its a bit odd that we depend on header here but dont get the entire header anymore

Reply:
Break it up how?

Reply:
as in remove the notion of header because we dont get the complete header anymore

Reply:
Doing a quick grok on `main`, it seems that ~80% of calls to `NewContext` pass in an empty header anyway. There are a few calls which pass in a non-empty header. What do you propose we do with those calls?

Reply:
remove header and only use WithCometInfo, WithHeaderInfo and WithBlockHeader to get the info into the header?

Reply:
ahhh yeah, we can do that!

Reply:
lets do this in a followup actually and get this merged, you can open against this but lets keep current state so we can merge faster